src/sentry/tasks/unmerge.py

```diff
@@ -435,60 +436,64 @@ def lock_hashes(project_id, source_id, fingerprints):
     return [h.hash for h in eligible_hashes]
 
 
-def unlock_hashes(project_id, fingerprints):
+def unlock_hashes(project_id, locked_primary_hashes):
     GroupHash.objects.filter(
-        project_id=project_id, hash__in=fingerprints, state=GroupHash.State.LOCKED_IN_MIGRATION
+        project_id=project_id,
+        hash__in=locked_primary_hashes,
+        state=GroupHash.State.LOCKED_IN_MIGRATION,
     ).update(state=GroupHash.State.UNLOCKED)
 
 
 @instrumented_task(name="sentry.tasks.unmerge", queue="unmerge")
-def unmerge(
-    project_id,
-    source_id,
-    destination_id,
-    fingerprints,
-    actor_id,
-    last_event=None,
-    batch_size=500,
-    source_fields_reset=False,
-    eventstream_state=None,
-):
-    source = Group.objects.get(project_id=project_id, id=source_id)
+def unmerge(*posargs, **kwargs):
+    args = UnmergeArgsBase.parse_arguments(*posargs, **kwargs)
+
+    source = Group.objects.get(project_id=args.project_id, id=args.source_id)
 
     caches = get_caches()
 
-    project = caches["Project"](project_id)
+    project = caches["Project"](args.project_id)
 
     # On the first iteration of this loop, we clear out all of the
     # denormalizations from the source group so that we can have a clean slate
     # for the new, repaired data.
-    if last_event is None:
-        fingerprints = lock_hashes(project_id, source_id, fingerprints)
+    if isinstance(args, InitialUnmergeArgs):
+        locked_primary_hashes = lock_hashes(
+            args.project_id, args.source_id, args.replacement.primary_hashes_to_lock
+        )
         truncate_denormalizations(project, source)
+        last_event = None
+    else:
+        last_event = args.last_event
+        locked_primary_hashes = args.locked_primary_hashes
 
     last_event, events = celery_run_batch_query(
-        filter=eventstore.Filter(project_ids=[project_id], group_ids=[source.id]),
-        batch_size=batch_size,
+        filter=eventstore.Filter(project_ids=[args.project_id], group_ids=[source.id]),
+        batch_size=args.batch_size,
         state=last_event,
         referrer="unmerge",
     )
 
     # If there are no more events to process, we're done with the migration.
     if not events:
-        unlock_hashes(project_id, fingerprints)
-        logger.warning("Unmerge complete (eventstream state: %s)", eventstream_state)
-        if eventstream_state:
-            eventstream.end_unmerge(eventstream_state)
-
-        return destination_id
+        unlock_hashes(args.project_id, locked_primary_hashes)
+        if isinstance(args, SuccessiveUnmergeArgs):
+            for unmerge_key, (group_id, eventstream_state) in args.destinations.items():
+                logger.warning("Unmerge complete (eventstream state: %s)", eventstream_state)
```

Comment:
Why is this warning now only logged when we're in the "successive" state?

Reply:
I'll change this to a better API. `InitialUnmergeArgs` should not expose/hold eventstream_state, as such it would not be available to pass into `stop_snuba_replacement`.